sensing/autoware_cuda_pointcloud_preprocessor/src/cuda_downsample_filter/cuda_voxel_grid_downsample_filter.cu

```diff
@@ -0,0 +1,471 @@
+// Copyright 2025 TIER IV, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "autoware/cuda_downsample_filter/cuda_voxel_grid_downsample_filter.hpp"
+#include "autoware/cuda_downsample_filter/thrust_custom_allocator.hpp"
+
+#include <sensor_msgs/msg/detail/point_field__struct.hpp>
+
+#include <thrust/adjacent_difference.h>
+#include <thrust/device_ptr.h>
+#include <thrust/device_vector.h>
+#include <thrust/extrema.h>
+#include <thrust/scan.h>
+#include <thrust/sort.h>
+#include <thrust/unique.h>
+
+#include <cmath>
+#include <cstdint>
+#include <memory>
+#include <stdexcept>
+#include <type_traits>
+
+namespace autoware::cuda_downsample_filter
+{
+namespace
+{
+template <typename T>
+__device__ T * getElementPointer(
+  uint8_t * data, const size_t point_index, const size_t point_step, const size_t offset)
+{
+  return reinterpret_cast<T *>(data + point_index * point_step + offset);
+}
+
+template <typename T>
+__device__ const T getElementValue(
```

Comment:
I am not sure how it works in cuda, but in c++ this looks like aliasing. Can you not assume that the input point type is the one defined in this package? That way you can reduce a lot of clutter IMO

Reply:
Yes, this is just aliasing. Since this kind of raw pointer arithmetic easily causes bugs, I chose to use aliasing with template typename. Besides, as handled in 0c949b35d, the intensity field may have a different data type, so I chose this design to keep availability as much as possible, even though the result becomes clutter :P